src/Network/Transport/TCP.hs

```diff
@@ -1951,9 +1955,20 @@ findRemoteEndPoint ourEndPoint theirAddress findOrigin mtimer = go
         const $ readMVar mv
 
 -- | Send a payload over a heavyweight connection (thread safe)
+--
+-- async shields non-atomic sendMany call from outer non-IOExceptions
+-- (like ProcessLinkException). remoteSendLock protects the socket
+-- from the concurrent threads usage as well as from the usage after
+-- an IOException.
 sendOn :: ValidRemoteEndPointState -> [ByteString] -> IO ()
-sendOn vst bs = withMVar (remoteSendLock vst) $ \() ->
-  sendMany (remoteSocket vst) bs
+sendOn vst bs = wait =<< async
+  (mask $ \restore -> do
+    is_good <- takeMVar (remoteSendLock vst)
+    when (is_good) $
+      restore (sendMany (remoteSocket vst) bs) `catch` \ex -> do
+        putMVar (remoteSendLock vst) False
+        throwIO (ex :: IOException)
```

Comment:
I don't see how this protects against an asynchronous exception as the documentation comment of `sendOn` claims.

Reply:
I mean, `ProcessLinkException` could still cause incomplete messages to be send.

Reply:
The protection against `ProcessLinkException`s is a bit above - at the async call which runs all this code in the new thread which is not targeted by such kind of exceptions, so it won't be interrupted.

Reply:
`async` looks fine for that purpose.

How about
```Haskell
    when (is_good) $
-     restore (sendMany (remoteSocket vst) bs) `catch` \ex -> do
+     restore (sendMany (remoteSocket vst) bs) `onException`
        putMVar (remoteSendLock vst) False
-       throwIO (ex :: IOException)
```
?

That is, why handle only IOExceptions instead of any synchronous exceptions?

Reply:
Well, the remote endpoint is failed only on IOExceptions, isn't it? If we do it this way, the sendOn may stop working (on some non-IOException) and we will never know why... (It will be very hard to debug it also.)

Reply:
> Well, the remote endpoint is failed only on IOExceptions, isn't it?
I've seen no documentation guaranteeing this.

Any exception can cause the message to be partially sent. From this perspective, it doesn't makes sense to allow subsequent calls to write to the thread.

> If we do it this way, the sendOn may stop working (on some non-IOException) and we will never know why...

Fair point. But writing to the socket after a partial message is not much easier to spot. How about `remoteSendLock vst` stores the reason why no more writes are allowed? Instead of making `sendOn` a noop, it could throw an exception in these cases.

Reply:
Interesting idea. Let me play with it...

Reply:
How about:

```haskell
  (mask $ \restore -> do
    is_good <- takeMVar (remoteSendLock vst)
    when (is_good) $
      restore (sendMany (remoteSocket vst) bs) `onException`
        putMVar (remoteSendLock vst) False
    putMVar (remoteSendLock vst) is_good
    when (not is_good) $
      throwIO $ userError "sendOn: invalid socket")
```

?

Reply:
I had something similar in mind. It would imply changing the type of `remoteSendLock`.

```haskell
  (mask $ \restore -> do
    is_good <- takeMVar (remoteSendLock vst)
    when (isNothing is_good) $
      restore (sendMany (remoteSocket vst) bs) `catch` \e -> do
        putMVar (remoteSendLock vst) (e :: SomeException)
        throwIO e
    putMVar (remoteSendLock vst) is_good
    case is_good of
      Nothing -> return ()
      Just e -> throwIO $ userError $ "sendOn failed earlier with: " ++ show e
```

Reply:
Take a look at the updated patch at https://github.com/haskell-distributed/network-transport-tcp/pull/86/commits/402504c894481c86267c907dc1931efd4b1e1d96.